src/client/wire_protocol/header.rs

```diff
@@ -0,0 +1,192 @@
+///                 should be `0` in a request from the client.
+/// `op_code`     - Identifies which type of message is being sent.
+// #[derive(Clone)]
+pub struct Header {
+    pub message_length: i32,
+    request_id: i32,
+    response_to: i32,
+    pub op_code: OpCode,
+}
+
+impl Header {
+    /// Constructs a new Header.
+    ///
+    /// # Arguments
+    ///
+    /// `message_length` - The length of the message in bytes.
+    /// `request_id` - Identifier for the request, or `0` if the the message
+    ///                is a response.
+    /// `response_to` - Identifies which request the message is in response to,
+    ///                or `0` if the the message is a request.
+    /// `op_code`    - Identifies which type of message is being sent.
+    ///
+    /// # Return value
+    ///
+    /// Returns the newly-created Header.
+    pub fn new(message_length: i32, request_id: i32, response_to: i32,
+           op_code: OpCode) -> Header {
+        Header { message_length: message_length, request_id: request_id,
+                 response_to: response_to, op_code: op_code }
+    }
+
+    /// Construcs a new Header for a request.
+    ///
+    /// # Arguments
+    ///
+    /// `message_length` - The length of the message in bytes.
+    /// `request_id` - Identifier for the request, or `0` if the the message
+    ///                is a response.
+    /// `op_code`    - Identifies which type of message is being sent.
+    ///
+    /// # Return value
+    ///
+    /// Returns a new Header with `response_to` set to 0.
+    fn new_request(message_length: i32, request_id: i32,
+                   op_code: OpCode) -> Header {
+        Header::new(message_length, request_id, 0, op_code)
+    }
+
+    fn new_reply(message_length: i32, response_to: i32,
+                 op_code: OpCode) -> Header {
+        Header::new(message_length, 0, response_to, op_code)
+    }
+
+    /// Constructs a new Header for a query.
+    ///
+    /// # Arguments
+    ///
+    /// `message_length` - The length of the message in bytes.
+    /// `request_id` - Identifier for the request, or `0` if the the message
+    ///                is a response.
+    /// # Return value
+    ///
+    /// Returns a new Header with `response_to` set to 0 and `op_code`
+    /// set to `OpQuery`.
+    pub fn new_query(message_length: i32, request_id: i32) -> Header {
+        Header::new_request(message_length, request_id, OpCode::OpQuery)
+    }
+
+
+    /// Writes the serialized Header to a buffer.
+    ///
+    /// # Arguments
+    ///
+    /// `buffer` - The buffer to write to.
+    ///
+    /// # Return value
+    ///
+    /// Returns nothing on success, or an error string on failure.
+    pub fn write(&self, buffer: &mut Write) -> Result<(), String> {
+        let _ = match buffer.write_i32::<LittleEndian>(self.message_length) {
+            Ok(_) => (),
+            Err(_) => return Err("Unable to write message_length".to_owned())
+        };
```

Comment:
Yeah, I'm talking about extraneous underscore matches. Everything except the error will be thrown out, so you can do ```if buffer.write_i32().is_err() { return Err("...") }```.